### PR TITLE
50 run tests requiring a bitbucket server in GitHub actions

### PR DESCRIPTION
**Pull request by jesperes**
This is a rather large PR which adds support for running the PropEr tests in `prop_api.erl` and `prop_yml.erl`. These tests require a running Bitbucket Server instance, which is the reason these have not been run before. This also caused problems because PRs opened against the GitHub repo were not being tested properly. (These tests were only run internally at Klarna.)

To spin up a Bitbucket Server instance, the `build.yml` workflow has been modified to start a Bitbucket service using the `atlassian/bitbucket-server:7.17.1` Docker image. It uses a "time-bomb" license retrieved from [here](https://developer.atlassian.com/platform/marketplace/timebomb-licenses-for-testing-server-apps/) which will automatically expire after three hours.

There is also a [docker-compose](https://docs.docker.com/compose/) setup in `scripts/pre-test-hook.sh` which is invoked automatically when running `rebar3 proper` locally. (Currently, to run these tests locally, you must use the `run-tests-with-bb.sh` script, but I hope to fix so that won't be necessary, and users should be able to just run `rebar3 proper` to test things, and things should "just work".

### CODE REVIEW

**Comment by jesperes** [comment]
Caveat: the workzone and hook tests will not be run, since these require additional plugins installed in Bitbucket.

File: test/bec_test_utils.erl [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+                          Password,
+                          maps:get(host, Map),
+                          maps:get(port, Map),
+                          ProjectKey,
+                          RepoSlug]),
+
+  LocalRepo = "/tmp/bec-test" ++ integer_to_list(rand:uniform(1 bsl 127)),
+  ok = cmd("mkdir -p ~s && cd ~s && "
+           "git init . && "
+           "git config user.email $USER@$HOSTNAME && "
```
Comment (mikpe) [inline test/bec_test_utils.erl:42]: nit: shouldn't these be hard-coded to something innocous like `jane.doe@example.com`?

File: test/bec_test_utils.erl [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+    ok = bitbucket:create_group(TeamA),
+    ok = bitbucket:create_group(TeamB),
+    ok = bitbucket:create_user(UserA, UserEmailA, UserDisplayNameA, <<"password">>),
+    ok = bitbucket:create_user(UserB, UserEmailB, UserDisplayNameB, <<"password">>),
+    ok = bitbucket:add_user_to_groups(UserA, [TeamA]),
+    ok = bitbucket:add_user_to_groups(UserB, [TeamB]),
+    init_repo()
+  catch C:T:St ->
+      %% If the setup function fails, the teardown function is not
+      %% executed and we risk leaving a running bitbucket instance in a
```
Comment (mikpe) [inline test/bec_test_utils.erl:82]: nit: _an_ inconsistent state

File: test/bec_test_utils.erl [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+        , deinit_bitbucket/0
+        , init_logging/0
+        , is_wz_supported/0
+        ]).
+
+-compile([{parse_transform, lager_transform}]).
+
+cmd(Fmt, Args) ->
+  Cmd = lists:flatten(io_lib:format(Fmt, Args)),
+  %% TODO log command output somewhere
```
Comment (mikpe) [inline test/bec_test_utils.erl:14]: you're enabling `lager` so why not that?

File: test/prop_api.erl [hunk 4]
```diff
@@ -423,18 +442,25 @@ setup() ->
   application:set_env(bec, bitbucket_username, Username),
   application:set_env(bec, bitbucket_password, Password),
   {ok, Started} = application:ensure_all_started(bec),
-  #{started => Started}.
+  bec_test_utils:init_bitbucket(),
+  bec_test_utils:init_logging(),
```
Comment (mikpe) [inline test/prop_api.erl:446]: I wonder if logging shouldn't be enabled first, and perhaps even be baked into `init_bitbucket/0`.

**Review by mikpe — APPROVED** [review]

File: test/bec_test_utils.erl (hunk 1 above)
Comment (jesperes) [reply test/bec_test_utils.erl:42]: Perhaps, but the repo is extremely shortlived (it is deleted in the same shell command). It only exists in order to populate the repository with a branch/commit.

**Pull request merged**